### PR TITLE
gh-108740: Fix "make regen-all" race condition

### PR DESCRIPTION
**Pull request by vstinner**
Fix a race condition in "make regen-all". The deepfreeze.c source and files generated by Argument Clinic are now generated or updated before generating "global objects". Previously, some identifiers may miss depending on the order in which these files were generated.

* "make regen-global-objects": Make sure that deepfreeze.c is generated and up to date, and always run "make clinic".
* "make clinic" no longer runs generate_global_objects.py script anymore.
* "make regen-deepfreeze" now only updates deepfreeze.c, it doesn't build deepfreeze.o anymore.

* Issue: gh-108740

### CODE REVIEW

**Comment by vstinner** [comment]
You can test this fix by adding a new global identifier with this patch:

```diff
index c66a862341..4534768305 100644
--- a/Modules/_asynciomodule.c
+++ b/Modules/_asynciomodule.c
@@ -780,6 +780,7 @@ _asyncio.Future.__init__
 
     *
     loop: object = None
+    new_funky_identifier: object = None
 
 This class is *almost* compatible with concurrent.futures.Future.
 
```

Run ``make regen-global-objects``: it should add the new identifier to global objects:

```
$ grep 'ID(new_funky_identifier)' ./Include/internal/pycore_global*.h
./Include/internal/pycore_global_objects_fini_generated.h:    _PyStaticObject_CheckRefcnt((PyObject *)&_Py_ID(new_funky_identifier));
./Include/internal/pycore_global_strings.h:        STRUCT_FOR_ID(new_funky_identifier)
```

Then revert the local patch, and make run that ``make regen-global-objects`` removes it.

Moreover, ``make clean; make regen-global-objects`` should no longer remove identifiers of ``deepfreze.c`` depending in which order ``deepfreze.c`` was generated or if it was not generated.

**Comment by vstinner** [comment]
I removed the misleading messages displayed to the user to ask to run ``make regen-global-objects`` . If a PR doesn't run ``make regen-all``, the "Check if generated files are up to date" job will fail anyway (with this fix).

In general, ``make regen-all`` makes sure that **all** generated files are up to date.

**Comment by vstinner** [comment]
Maybe:

* ``make Python/deepfreeze/deepfreeze.c`` should depend on ``make regen-frozen`` which can update ``$(FROZEN_FILES_OUT)``
* ``make regen-global-objects`` should depend on ``make regen-cases`` (ex: ``Python/generated_cases.c`` contains ``_Py_ID(...)``)

**Comment by vstinner** [comment]
cc @erlend-aasland

**Comment by markshannon** [comment]
We are looking at removing deepfreeze altogether. https://github.com/python/cpython/pull/108722

**Comment by vstinner** [comment]
> We are looking at removing deepfreeze altogether. https://github.com/python/cpython/pull/108722

Good to know! Anyway, it's still worth it to fix Python 3.12 build system (which uses ``deepfreeze``).

And there is still a dependency issue between ``make clinic`` and ``make regen-global-objects``.

**Comment by vstinner** [comment]
I updated my PR to better explain the change, and I documented generated files in Doc/using/configure.rst.

@erlend-aasland: Would you be available to review my change?

File: Doc/using/configure.rst [hunk 1]
```diff
@@ -60,6 +60,29 @@ See also :pep:`7` "Style Guide for C Code" and :pep:`11` "CPython platform
...
+
+To reduce build dependencies, Python source code contains multiple generated
+files. Commands to regenerate all generated files::
+
+    make regen-all
+    make regen-stdlib-module-names
+    make regen-limited-abi
+    make regen-configure
+
+The ``Makefile.pre.in`` documents generated files, their inputs, and tools used
```
Comment (vstinner) [inline Doc/using/configure.rst]:
```suggestion
The ``Makefile.pre.in`` file documents generated files, their inputs, and tools used
```

**Comment by erlend-aasland** [comment]
> I updated my PR to better explain the change, and I documented generated files in Doc/using/configure.rst.
> 
> @erlend-aasland: Would you be available to review my change?

I'll have a look tomorrow!

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
@markshannon:

> We are looking at removing deepfreeze altogether. #108722

IIUC, it is only partially removed, and the infrastructure remains, so it would make sense to fix the race condition.

**Comment by vstinner** [comment]
> Revert "Use DEEPFREEZE_C in target rule as well"

What what the problem?

> Use $(DEEPFREEZE_C) 

I tried again your change :-)

**Comment by vstinner** [comment]
> Tests / Check if generated files are up to date (pull_request) Successful

Good :-)

**Comment by vstinner** [comment]
@erlend-aasland: You may want to review backports, I had conflicts.

* 3.12 branch: https://github.com/python/cpython/pull/109019
* 3.11 branch: https://github.com/python/cpython/pull/109021

**Comment by vstinner** [comment]
Thanks for the review @erlend-aasland, I merged my PR.

**Pull request merged**